solidity/contracts/ManagedGrantFactory.sol

```diff
@@ -0,0 +1,96 @@
+pragma solidity ^0.5.4;
+
+import "openzeppelin-solidity/contracts/token/ERC20/SafeERC20.sol";
+
+import "./utils/BytesLib.sol";
+
+import "./KeepToken.sol";
+import "./ManagedGrant.sol";
+import "./TokenGrant.sol";
+import "./GrantStakingPolicy.sol";
+
+contract ManagedGrantFactory {
+    using SafeERC20 for KeepToken;
+    using BytesLib for bytes;
+    using BytesLib for address;
+
+    KeepToken public token;
+    mapping(address => uint256) public grantFundingPool;
+    TokenGrant public tokenGrant;
+    GrantStakingPolicy nonRevocableStakingPolicy;
+    GrantStakingPolicy revocableStakingPolicy;
+
+    constructor(
+        address _tokenAddress,
+        address _tokenGrant,
+        address _nonRevocableStakingPolicy,
+        address _revocableStakingPolicy
+    ) public {
+        token = KeepToken(_tokenAddress);
+        tokenGrant = TokenGrant(_tokenGrant);
+        nonRevocableStakingPolicy = GrantStakingPolicy(_nonRevocableStakingPolicy);
+        revocableStakingPolicy = GrantStakingPolicy(_revocableStakingPolicy);
+    }
+
+    function receiveApproval(
+        address _from,
+        uint256 _amount,
+        address _token,
+        bytes memory _extraData
```

Comment:
I think we should mirror the API of `TokenGrant`and `createGrant` in the same transaction as `receiveApproval` using `_extraData` to pass all the details.

It seems more convenient - we can create grants from dApp without setting up the factory first and we don't need to think about functionalities like withdrawal from the factory. It's also consistent with what we have in `TokenGrant.sol` and cheaper.

Reply:
If we do this I'd like to request a change to the `TokenGrant` API (and other applicable APIs like `TokenStaking`) so that it could use [`abi.decode`](https://solidity.readthedocs.io/en/v0.5.10/units-and-global-variables.html#abi-encoding-and-decoding-functions) directly instead of manually de/constructing the data.

If the API here is identical to `TokenGrant.receiveApproval` we still can't pass `_extraData` straight through, as we need to get the managed grant address etc. and then change the relevant fields. `abi.decode` adds some slight overhead by requiring unnecessary empty bytes to be transmitted, but these functions aren't in any hot or even warm paths, and the resulting code would be significantly simpler.

Reply:
Another reason we can't really mirror the API is that we need to return the `ManagedGrant` address and `approveAndCall` only returns `bool`. We'd need to return the address in an event which then necessitates complex and somewhat fragile workarounds if another contract wants to perform those actions (like how the `ManagedGrantFactory` itself uses `numGrants()` to get the ID of the next grant, which I hate but can't get around).

One possible workaround would be to have _both_ the specific function which requires a setup and returns useful values, _and_ `receiveApproval` which can be called through the token but only gives info in events. In this case `receiveApproval` would be a relatively thin wrapper around `createGrant/stakeTokens/etc` and both use-cases would be accommodated.

Reply:
I made a separate PR for the API at #1592

Reply:
Yeah, I don't like `approve` because of the re-approval attack but I have no better idea on how to solve it than what you proposed.